core/shared/src/main/scala/matryoshka/Recursive.scala

```diff
@@ -101,13 +115,37 @@ trait Recursive[T] extends Based[T] {
       : A =
     gcata[(B, ?), A](t)(distZygo(f), g)
 
+  def zygoM[A, B, M[_]: Monad]
+    (t: T)
+    (f: AlgebraM[M, Base, B], g: GAlgebraM[(B, ?), M, Base, A])
+    (implicit BT: Traverse[Base])
+      : M[A] = {
+    val k = distZygoM(f, distApplicative[Base, M])
+
+    def loop[W[_]: Comonad : Traverse]
+      (t: T)
+      (g: GAlgebraM[W, M, Base, A], k: DistributiveLaw[Base, (M ∘ W)#λ])
+        : M[W[Base[W[A]]]] =
+      project(t).traverse(loop(_)(g, k) >>=
+        (_.traverse(g) ∘ (_.cojoin))) ∘ (_ ∘ (_.point[M])) >>= (k(_))
+
+    loop[(B, ?)](t)(g, k) ∘ (_.copoint) >>= g
+  }
+
   def elgotZygo[A, B]
     (t: T)
     (f: Algebra[Base, B], g: ElgotAlgebra[(B, ?), Base, A])
     (implicit BF: Functor[Base])
       : A =
     elgotCata[(B, ?), A](t)(distZygo(f), g)
 
+  def elgotZygoM[A, B, M[_]: Monad]
+    (t: T)
+    (f: AlgebraM[M, Base, B], g: ElgotAlgebraM[(B, ?), M, Base, A])
+    (implicit BT: Traverse[Base])
+      : M[A] =
+    elgotCataM[A, B, (B, ?), M](t)(distZygoM(f, distApplicative[Base, M]), g)
```

Comment:
Yeah, it’s interesting that you can use `distZygoM` here in `elgotCataM`, but not in `gCataM` above.